edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,244 @@
+            return snapshot.TryGetValue(id, out ServiceIdentityTreeNode treeNode)
+                    ? Option.Some(treeNode.Identity)
+                    : Option.None<ServiceIdentity>();
+        }
+
+        public IList<string> GetAllIds()
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.Select(kvp => kvp.Value.Identity.Id).ToList();
+        }
+
+        public bool TryGetAuthChain(string id, out string authChain)
+        {
+            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+            authChain = string.Empty;
+
+            // Auth-chain for root
+            if (this.rootDeviceId == id)
+            {
+                authChain = this.rootDeviceId;
+                return true;
+            }
+
+            // Auth-chain for a child somewhere in the tree
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            if (snapshot.TryGetValue(id, out ServiceIdentityTreeNode treeNode))
+            {
+                // Auth chain starts with the originating device
+                authChain = treeNode.Identity.Id;
+
+                // Walk the parents up to the root
+                while (treeNode.Parent.HasValue)
+                {
+                    // Accumulate the auth chain per parent visited in the format:
+                    // "leaf1;edge1;edge2;...;rootEdge"
+                    ServiceIdentityTreeNode parent = treeNode.Parent.Expect(() => new InvalidOperationException());
+                    authChain += ";" + parent.Identity.Id;
+
+                    if (parent.Identity.DeviceId == this.rootDeviceId)
+                    {
+                        // We've reached the root device, we're done
+                        return true;
+                    }
+
+                    // Move on to the next parent
+                    treeNode = parent;
+                }
+
+                // If we walked the parent chain up to a dead-end, i.e. topmost node does not
+                // match tree root device ID, then it means this subtree is dangling.
+                // This is a possible scenario, we just treat it as if there's no auth chain.
+            }
+
+            // No valid auth chain
+            return false;
+        }
+
+        void InsertModuleIdentity(IDictionary<string, ServiceIdentityTreeNode> nodes, ServiceIdentity module)
+        {
+            var newNode = new ServiceIdentityTreeNode(module);
+            nodes.Add(module.Id, newNode);
+
+            if (nodes.TryGetValue(module.DeviceId, out ServiceIdentityTreeNode parentDeviceNode))
+            {
+                // Hook up the module to the parent device
+                newNode.Parent = Option.Some(parentDeviceNode);
+                parentDeviceNode.Children.Add(newNode);
+            }
+            else
+            {
+                // It's okay if we can't find the parent device, it's possible that the identity
```

Comment:
This happens if the child connects before we have had a chance to sync the tree?

Reply:
My concern here is - we are not actively doing anything to complete the tree.. in that case GetAuthChain will not be able to create the chain and we will not do anything about it.

Reply:
It feels like we should either fill up the tree up-front (and in updates), or fill it up on demand on GetAuthChain

Reply:
We discussed offline and when loading from storage there is no order so that would be the case when children are added before parent